File: mailblastr-python/mailblastr/webhooks.py
"""Webhooks resource + local Svix-style signature verification.

``Webhooks.verify(payload, headers, secret)`` (or the module-level
``verify_webhook_signature``) is a pure local computation using hmac/hashlib —
it makes no HTTP request. It mirrors the backend signing scheme:
``"{svix-id}.{svix-timestamp}.{body}"`` → base64 HMAC-SHA256, tagged ``v1,``.
"""

import base64
import binascii
import hashlib
import hmac
import math
import time
from typing import List, TypedDict

from . import http_client
from .http_client import paginate, path_escape as _e


class CreateParams(TypedDict, total=False):
    endpoint: str  # required
    events: List[str]  # required, e.g. ['email.delivered', 'email.bounced']
    secret: str  # optional caller-supplied signing secret (else generated, returned once)


class UpdateParams(TypedDict, total=False):
    endpoint: str
    events: List[str]
    status: str  # 'enabled' | 'disabled'


def _header_pairs(headers):
    """Best-effort ``(name, value)`` pairs out of ANY header container.

    ``items()`` first, because that is the one accessor every header type in the
    Python web stack agrees on: dict, ``http.client.HTTPMessage``, Django's
    ``request.headers``, Starlette/FastAPI's ``Headers`` — and Flask's, which is
    the one that made this necessary. Werkzeug's ``Headers.__iter__`` yields
    ``(key, value)`` TUPLES rather than keys, so iterating it and treating each
    element as a name matched nothing and ``Webhooks.verify(body,
    request.headers, secret)`` — the call this package's own README shows —
    answered ``missing_headers`` for every genuinely signed delivery.
    """
    if headers is None:
        return []
    items = getattr(headers, "items", None)
    if callable(items):
        try:
            return list(items())
        except Exception:  # pragma: no cover - exotic container
            pass
    try:
        entries = list(headers)
    except TypeError:
        return []
    pairs = []
    for entry in entries:
        # A pair-yielding container (werkzeug) vs. a mapping that yields keys.
        if isinstance(entry, (tuple, list)) and len(entry) == 2:
            pairs.append((entry[0], entry[1]))
            continue
        try:
            pairs.append((entry, headers[entry]))
        except (KeyError, IndexError, TypeError, AttributeError):
            continue
    return pairs


def _read_header(headers, name):
    """Case-insensitively read a single header value (first if it's a list)."""
    lower = name.lower()
    for key, value in _header_pairs(headers):
        if str(key).lower() == lower:
            if isinstance(value, (list, tuple)):
                return value[0] if value else None
            return value
    return None


# Base64 alphabet as BYTE values, plus the URL-safe spellings of the last two
# characters. Bytes, not characters, because rule 5 in `_b64_lenient` reduces the
# suffix to bytes before anything in it is looked up in this table.
_B64_ALPHABET = frozenset(
    b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/-_"
)

# `-`/`_` are the URL-safe spellings of `+`/`/`: TRANSLATED, never dropped (rule 3).
_B64_URLSAFE = bytes.maketrans(b"-_", b"+/")


def _b64_lenient(text):
    """Decode base64 the way the signer does, not the way Python does.

    The backend derives its key with Node's ``Buffer.from(suffix, 'base64')``
    (lib/crypto.ts ``secretToKey``), and every SDK has to reproduce that byte for
    byte: a key that differs from the signer's does not fail loudly, it reports
    ``no_match``, so a correctly configured endpoint silently treats every
    genuine delivery as forged. Node's rules are NOT the RFC's — following the
    RFC is exactly how this shipped broken twice — and all five of these matter:

    1. ``=`` TERMINATES the input. Everything from the first ``=`` onward is
       DISCARDED; it is not "padding to be stripped". ``"YWJj====ZA"`` is
       ``b"abc"``, not ``b"abcd"``, and ``"Y=WJj"`` decodes to NOTHING (which
       sends ``_secret_to_key`` down its raw-bytes fallback). Stripping ``=``
       and decoding the rest was this function's own residual divergence — the
       one 5.0.0 propagated to five other SDKs by porting it.
    2. Characters outside the alphabet are SKIPPED, never fatal: whitespace,
       punctuation and non-ASCII are ignored, so ``"YW!Jj"`` is ``b"abc"``.
    3. ``-`` and ``_`` are the URL-safe spellings of ``+`` and ``/`` and must be
       TRANSLATED, not dropped.
    4. A trailing group of ONE character carries no whole byte and contributes
       nothing; 2 chars give 1 byte, 3 give 2.
    5. The unit Node reads is the LOW 8 BITS OF EACH UTF-16 CODE UNIT, not the
       codepoint and not the UTF-8 bytes: `Buffer.from(str, 'base64')` masks
       every code unit with ``0xFF`` before the table lookup. So ``"YWŁj"`` is
       ``b"abc"`` mangled to ``616023`` (U+0141 & 0xFF = 0x41 ``A``), U+013D
       masks to 0x3D and TERMINATES the input under rule 1, and an astral
       character contributes its two SURROGATE halves' low bytes — ``"𝑁"`` is
       0xD835/0xDC41 → ``"5A"`` → the single byte 0xE4, never its UTF-8 bytes.
       This is
       why it must run BEFORE rules 1-4: split the raw string on ``=`` and
       U+013D never terminates anything. Invisible to ASCII tests, which is how
       all nine SDKs scored a clean sweep on the old corpus while diverging on
       every secret carrying a codepoint above 0xFF.

    ``base64.b64decode`` obeys none of this — it raises ``binascii.Error`` on a
    length that is not a multiple of four, silently discards ``-``/``_``, and
    happily decodes past an interior ``=``. So it is used only for the final
    fixed-up group decode; the truncation, the alphabet filter and the URL-safe
    translation are done here, because that is where every language's built-in
    parts ways with Node. A caller-supplied secret (the ``secret`` field on
    POST /webhooks, which the backend accepts verbatim with no shape
    validation) can be any of these shapes.
    """
    # Rule 5 first: UTF-16LE lays out two bytes per code unit little-endian, so
    # `[0::2]` IS the `& 0xFF` mask. `surrogatepass` is mandatory — without it an
    # astral character (or a lone surrogate a caller pasted in) raises instead of
    # yielding the two halves Node sees, and `text.encode("utf-8")` here would
    # give U+1D441 three bytes of entirely different values.
    masked = text.encode("utf-16-le", "surrogatepass")[0::2]
    head = masked.split(b"=", 1)[0]
    cleaned = bytes(b for b in head if b in _B64_ALPHABET).translate(_B64_URLSAFE)
    remainder = len(cleaned) % 4
    if remainder == 1:
        cleaned = cleaned[:-1]  # a single leftover character encodes no byte
    elif remainder:
        cleaned += b"=" * (4 - remainder)
    try:
        return base64.b64decode(cleaned)
    except (ValueError, binascii.Error):  # pragma: no cover - defensive
        return b""


def _secret_to_key(secret):
    """Derive the HMAC key from a ``whsec_``-prefixed secret (base64-decode the
    suffix); a secret without the prefix is used as raw UTF-8 bytes.

    The fallback shape is as load-bearing as the decoder and is easy to get
    subtly wrong: when the suffix decodes to ZERO bytes — ``whsec_``, ``whsec_=``,
    ``whsec_Y``, ``whsec_!!!!`` — the server keys off the WHOLE secret including
    the ``whsec_`` prefix, not the suffix and not an empty key. Anything else
    here fails exactly as invisibly as a wrong decoder.
    """
    s = str(secret or "")
    if s.startswith("whsec_"):
        decoded = _b64_lenient(s[len("whsec_"):])
        if decoded:
            return decoded
    return s.encode("utf-8")


def verify_webhook_signature(payload, headers, secret, tolerance_sec=300):
    """Verify a webhook delivery's Svix-style signature against your endpoint's
    signing secret.

    ``payload`` MUST be the exact raw request body string the server sent (do
    not re-serialize parsed JSON — whitespace differences break the signature).
    ``headers`` accepts the svix-id / svix-timestamp / svix-signature headers
    (read case-insensitively); a header may carry multiple space-separated
    signatures, any one matching makes the delivery valid.

    ``tolerance_sec`` is the max allowed clock skew (default 300; pass 0 to
    skip the freshness check).

    Returns ``{"valid": True}`` or ``{"valid": False, "reason": ...}``.
    """
    if isinstance(payload, bytes):
        payload = payload.decode("utf-8")
    sig_id = _read_header(headers, "svix-id")
    timestamp = _read_header(headers, "svix-timestamp")
    sig_header = _read_header(headers, "svix-signature")
    if not sig_id or not timestamp or not sig_header:
        return {"valid": False, "reason": "missing_headers"}
    if not secret:
        return {"valid": False, "reason": "missing_secret"}

    # Optional timestamp freshness check (default 5-minute tolerance; 0 disables).
    if tolerance_sec and tolerance_sec > 0:
        try:
            ts = float(timestamp)
        except (TypeError, ValueError):
            return {"valid": False, "reason": "invalid_timestamp"}
        if not math.isfinite(ts):
            return {"valid": False, "reason": "invalid_timestamp"}
        if abs(int(time.time()) - ts) > tolerance_sec:
            return {"valid": False, "reason": "timestamp_out_of_tolerance"}

    signed = f"{sig_id}.{timestamp}.{payload}"
    expected = base64.b64encode(
        hmac.new(_secret_to_key(secret), signed.encode("utf-8"), hashlib.sha256).digest()
    ).decode("ascii")

    # The header may contain multiple space-separated `v1,<sig>` entries; any match wins.
    for part in str(sig_header).split(" "):
        part = part.strip()
        if not part:
            continue
        sig = part[len("v1,"):] if part.startswith("v1,") else part
        if hmac.compare_digest(sig, expected):
            return {"valid": True}
    return {"valid": False, "reason": "no_match"}


class Webhooks:
    """``mailblastr.Webhooks`` — the /webhooks endpoints."""

    CreateParams = CreateParams
    UpdateParams = UpdateParams

    @classmethod
    def create(cls, params):
        """Create a webhook. The signing secret is shown ONCE, only here.
        POST /webhooks"""
        return http_client.request("POST", "/webhooks", params)

    @classmethod
    def get(cls, webhook_id):
        """Retrieve a webhook. GET /webhooks/:id"""
        return http_client.request("GET", f"/webhooks/{_e(webhook_id)}")

    @classmethod
    def list(cls, params=None):
        """List webhooks. GET /webhooks"""
        return http_client.request("GET", f"/webhooks{paginate(params)}")

    @classmethod
    def update(cls, webhook_id, params):
        """Update a webhook. PATCH /webhooks/:id"""
        return http_client.request("PATCH", f"/webhooks/{_e(webhook_id)}", params)

    @classmethod
    def rotate(cls, webhook_id):
        """Rotate the signing secret. The new plaintext ``signing_secret`` is
        returned ONCE; the old secret stops verifying immediately.
        POST /webhooks/:id/rotate"""
        return http_client.request("POST", f"/webhooks/{_e(webhook_id)}/rotate")

    @classmethod
    def test(cls, webhook_id):
        """Send a synchronous test delivery and return the endpoint's live
        result. POST /webhooks/:id/test

        A FAILED delivery is still HTTP 200, so this does NOT raise when your
        endpoint rejects the test. The outcome is the ``ok`` key::

            {"object": "webhook_test", "id": "<id>",
             "ok": True,  "status": 200}                 # endpoint accepted it
            {"object": "webhook_test", "id": "<id>",
             "ok": False, "error": "lookup_failed"}      # it did not

            result = mailblastr.Webhooks.test(webhook_id)
            if not result["ok"]:
                print("test delivery failed:", result.get("error"))

        ``status`` is your endpoint's HTTP status when it responded at all;
        ``error`` says why the delivery failed (e.g. ``"lookup_failed"``,
        ``"webhook missing or disabled"``). It is a single attempt — no
        retries are scheduled.
        """
        return http_client.request("POST", f"/webhooks/{_e(webhook_id)}/test")

    @classmethod
    def remove(cls, webhook_id):
        """Delete a webhook. DELETE /webhooks/:id"""
        return http_client.request("DELETE", f"/webhooks/{_e(webhook_id)}")

    @classmethod
    def verify(cls, payload, headers, secret, tolerance_sec=300):
        """Verify a webhook delivery signature (pure local computation).
        See :func:`mailblastr.verify_webhook_signature`."""
        return verify_webhook_signature(payload, headers, secret, tolerance_sec)

File: mailblastr-java/src/main/java/com/mailblastr/resources/Webhooks.java
package com.mailblastr.resources;

import com.mailblastr.ListParams;
import com.mailblastr.MailblastrResponse;
import com.mailblastr.http.ApiClient;
import com.mailblastr.requests.CreateWebhookRequest;
import com.mailblastr.requests.UpdateWebhookRequest;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.util.Base64;
import java.util.Map;

/** Webhooks — {@code mailblastr.webhooks()} — including local Svix-style signature verification. */
public final class Webhooks extends Resource {
    public Webhooks(ApiClient api) { super(api); }

    /** Create a webhook. The signing secret is shown ONCE, only here. {@code POST /webhooks} */
    public MailblastrResponse create(CreateWebhookRequest request) {
        return api.request("POST", "/webhooks", request);
    }

    /** {@code GET /webhooks/:id} */
    public MailblastrResponse get(String id) {
        return api.request("GET", "/webhooks/" + enc(id));
    }

    /**
     * List webhooks. Unlike domains/api-keys/topics this route always applies a
     * limit — with no pagination params you get the first 20.
     * {@code GET /webhooks}
     */
    public MailblastrResponse list() { return list(null); }

    public MailblastrResponse list(ListParams params) {
        return api.request("GET", "/webhooks" + paginate(params));
    }

    /** Returns the slim ack {@code { object: 'webhook', id }}. {@code PATCH /webhooks/:id} */
    public MailblastrResponse update(String id, UpdateWebhookRequest request) {
        return api.request("PATCH", "/webhooks/" + enc(id), request);
    }

    /**
     * Rotate the signing secret — the new plaintext {@code signing_secret} is
     * returned ONCE; the old secret stops verifying immediately.
     * {@code POST /webhooks/:id/rotate}
     */
    public MailblastrResponse rotate(String id) {
        return api.request("POST", "/webhooks/" + enc(id) + "/rotate");
    }

    /**
     * Send a synchronous, single-attempt test delivery and return the
     * endpoint's live result.
     *
     * <p><strong>A failed delivery is still HTTP 200</strong> — this method
     * does not throw for one. Inspect {@code ok} on the response
     * ({@code res.getBoolean("ok")}) and read {@code error} when it is false.
     * {@code POST /webhooks/:id/test}
     */
    public MailblastrResponse test(String id) {
        return api.request("POST", "/webhooks/" + enc(id) + "/test");
    }

    /** {@code DELETE /webhooks/:id} */
    public MailblastrResponse remove(String id) {
        return api.request("DELETE", "/webhooks/" + enc(id));
    }

    /** Instance alias of {@link #verifyWebhookSignature(String, Map, String)}. */
    public VerifyWebhookResult verify(String payload, Map<String, String> headers, String secret) {
        return verifyWebhookSignature(payload, headers, secret);
    }

    /** Instance alias of {@link #verifyWebhookSignature(String, Map, String, long)}. */
    public VerifyWebhookResult verify(String payload, Map<String, String> headers, String secret, long toleranceSec) {
        return verifyWebhookSignature(payload, headers, secret, toleranceSec);
    }

    /** Verify with the default 5-minute timestamp tolerance. */
    public static VerifyWebhookResult verifyWebhookSignature(String payload, Map<String, String> headers, String secret) {
        return verifyWebhookSignature(payload, headers, secret, 300);
    }

    /**
     * Verify a webhook delivery's Svix-style signature against your endpoint's
     * signing secret. Mirrors the backend signing scheme:
     * {@code "<svix-id>.<svix-timestamp>.<body>"} → base64 HMAC-SHA256, tagged
     * {@code v1,}.
     *
     * <p>{@code payload} MUST be the exact raw request body string the server
     * sent — do not re-serialize parsed JSON (whitespace differences break the
     * signature). {@code headers} must contain the {@code svix-id},
     * {@code svix-timestamp} and {@code svix-signature} headers (read
     * case-insensitively). A signature header may carry multiple
     * space-separated signatures; any one matching makes the delivery valid.
     * A {@code whsec_}-prefixed secret is base64-decoded exactly as leniently
     * as the signer decodes it — unpadded, junk-bearing and URL-safe
     * ({@code -}/{@code _}) spellings all derive the same key, a {@code =}
     * truncates the suffix rather than padding it, and a non-ASCII character
     * decodes as the low byte of its UTF-16 code unit rather than being skipped
     * (see {@link #b64Lenient(String)}). A secret without the prefix, or one whose
     * suffix decodes to nothing, is used as the raw UTF-8 bytes of the WHOLE
     * secret string.
     *
     * <p>This is a pure local computation — it makes no HTTP request.
     *
     * @param toleranceSec max allowed clock skew in seconds (0 disables the check)
     */
    public static VerifyWebhookResult verifyWebhookSignature(
            String payload, Map<String, String> headers, String secret, long toleranceSec) {
        String id = readHeader(headers, "svix-id");
        String timestamp = readHeader(headers, "svix-timestamp");
        String sigHeader = readHeader(headers, "svix-signature");
        if (isBlank(id) || isBlank(timestamp) || isBlank(sigHeader)) {
            return new VerifyWebhookResult(false, "missing_headers");
        }
        if (isBlank(secret)) {
            return new VerifyWebhookResult(false, "missing_secret");
        }

        // Optional timestamp freshness check (default 5-minute tolerance; 0 disables).
        if (toleranceSec > 0) {
            long ts;
            try {
                ts = Long.parseLong(timestamp.trim());
            } catch (NumberFormatException e) {
                return new VerifyWebhookResult(false, "invalid_timestamp");
            }
            long skew = Math.abs(System.currentTimeMillis() / 1000L - ts);
            if (skew > toleranceSec) {
                return new VerifyWebhookResult(false, "timestamp_out_of_tolerance");
            }
        }

        String signed = id + "." + timestamp + "." + (payload == null ? "" : payload);
        String expected;
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secretToKey(secret), "HmacSHA256"));
            expected = Base64.getEncoder().encodeToString(mac.doFinal(signed.getBytes(StandardCharsets.UTF_8)));
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("HmacSHA256 unavailable", e);
        }

        // The header may contain multiple space-separated `v1,<sig>` entries; any match wins.
        for (String part : sigHeader.split(" ")) {
            String trimmed = part.trim();
            if (trimmed.isEmpty()) continue;
            String sig = trimmed.startsWith("v1,") ? trimmed.substring(3) : trimmed;
            if (MessageDigest.isEqual(
                    sig.getBytes(StandardCharsets.UTF_8),
                    expected.getBytes(StandardCharsets.UTF_8))) {
                return new VerifyWebhookResult(true, null);
            }
        }
        return new VerifyWebhookResult(false, "no_match");
    }

    /** Base64 alphabet, plus the URL-safe spellings of the last two characters. */
    private static final String B64_ALPHABET =
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/-_";

    /**
     * Decode base64 the way the signer does, not the way Java does.
     *
     * <p>The backend derives its key with Node's {@code Buffer.from(suffix,
     * 'base64')} (lib/crypto.ts {@code secretToKey}). That decoder is NOT the
     * RFC, and reading it as the RFC is precisely how this bug shipped twice.
     * Five of its behaviours matter, and each one has a conformance vector:
     *
     * <ol>
     *   <li>{@code =} TERMINATES the input — everything from the first
     *       {@code =} onward is DISCARDED. It is not "padding to be stripped":
     *       {@code "YWJj====ZA"} decodes to {@code "abc"}, never
     *       {@code "abcd"}, and {@code "Y=WJj"} decodes to nothing at all.</li>
     *   <li>Out-of-alphabet characters are SKIPPED, never fatal — whitespace,
     *       punctuation and non-ASCII alike ({@code "YW!Jj"} is {@code "abc"}).</li>
     *   <li>{@code -} and {@code _} are the URL-safe spellings of {@code +} and
     *       {@code /} and must be TRANSLATED, not dropped — dropping them
     *       silently reshapes the whole byte stream.</li>
     *   <li>A trailing group of ONE character carries no whole byte and is
     *       discarded (2 chars -&gt; 1 byte, 3 -&gt; 2, 4 -&gt; 3).</li>
     *   <li>The decoded UNIT is the LOW 8 BITS of each UTF-16 CODE UNIT, not the
     *       codepoint: Node masks every code unit with {@code 0xFF} before the
     *       table lookup, so a non-ASCII character is not junk to be skipped —
     *       it MASQUERADES as the ASCII character it aliases. {@code "YWŁj"}
     *       keys on {@code 616023} because U+0141 masks to {@code 0x41} 'A';
     *       {@code "YWJjĽZA"} keys on {@code 616263} because U+013D masks to
     *       {@code 0x3D} '=' and TERMINATES. Astral characters contribute their
     *       two SURROGATE halves, not their UTF-8 bytes: U+1D441 is
     *       {@code 0xD835}/{@code 0xDC41}, masking to '5' and 'A'.</li>
     * </ol>
     *
     * <p>Java agrees with none of these unaided. {@code Base64.getDecoder()}
     * throws on {@code -}/{@code _}, on any other out-of-alphabet character and
     * on a trailing 1-character group; {@code getMimeDecoder()} skips junk but
     * treats {@code =} as padding it may decode straight past, and both read the
     * character, never its low byte. So the mask, the cut at the first
     * {@code =}, the alphabet filter and the URL-safe translation are all done
     * HERE, by hand, before Java's decoder is allowed near the input.
     *
     * <p>Why this matters more than it looks: the {@code secret} field on
     * {@code POST /webhooks} is stored verbatim with no shape validation, so a
     * customer really can hold a secret of any shape above. A key that differs
     * from the signer's does not fail loudly — verification just returns
     * {@code no_match}, and a correctly configured endpoint silently treats
     * every genuine delivery as forged.
     */
    private static byte[] b64Lenient(String text) {
        StringBuilder cleaned = new StringBuilder(text.length());
        for (int i = 0; i < text.length(); i++) {
            // Rule 5, and it MUST come first: mask to the low 8 bits of the
            // UTF-16 code unit. `charAt` deliberately walks code UNITS, so an
            // astral character arrives as its two surrogates and contributes
            // both their low bytes — iterating codePoints would collapse it to
            // one wrong unit. Masking after the `=` cut or the alphabet filter
            // is the bug: U+013D would be skipped as junk instead of
            // terminating, and U+0141 dropped instead of aliasing 'A'.
            char c = (char) (text.charAt(i) & 0xFF);

            // Rule 1: `=` TERMINATES — the remainder of the suffix is gone, not
            // padding to be stripped. Filtering it away as junk and decoding on
            // is the shared root-cause bug every SDK carried.
            if (c == '=') break;

            // Rules 2 and 3: keep alphabet characters only, translating the
            // URL-safe spellings instead of discarding them.
            if (B64_ALPHABET.indexOf(c) < 0) continue;
            cleaned.append(c == '-' ? '+' : c == '_' ? '/' : c);
        }

        // Rule 4: a single leftover character encodes no whole byte; Node
        // discards it rather than failing the decode. Java's MIME decoder would
        // throw ("Last unit does not have enough valid bits") and lose the key.
        if (cleaned.length() % 4 == 1) cleaned.setLength(cleaned.length() - 1);
        try {
            // The MIME decoder, unlike getDecoder(), tolerates absent padding.
            return Base64.getMimeDecoder().decode(cleaned.toString());
        } catch (IllegalArgumentException ignored) { // defensive — input is pre-cleaned
            return new byte[0];
        }
    }

    /**
     * Derive the HMAC key from a secret, mirroring the server's
     * {@code secretToKey} (lib/crypto.ts) exactly.
     *
     * <p>A {@code whsec_}-prefixed secret keys on the leniently decoded suffix
     * (see {@link #b64Lenient(String)}). Everything else — no prefix, or a
     * suffix that decodes to ZERO bytes — keys on the raw UTF-8 bytes of the
     * WHOLE secret, {@code whsec_} prefix INCLUDED. That last clause is easy to
     * miss and easy to get subtly wrong (keying on the bare suffix, or on the
     * empty decode, both look reasonable); it is reached by every secret whose
     * suffix is one alphabet character, is empty, or begins with {@code =}, and
     * getting it wrong costs the customer every genuine delivery.
     *
     * <p>The fallback takes the secret's REAL UTF-8 bytes — the {@code 0xFF}
     * masking of {@link #b64Lenient(String)} belongs to the base64 table lookup
     * only, so {@code "whsec_š"} falls back to the 8 bytes of {@code "whsec_š"},
     * not to the masked {@code 0x61}.
     */
    private static byte[] secretToKey(String secret) {
        if (secret.startsWith("whsec_")) {
            byte[] decoded = b64Lenient(secret.substring("whsec_".length()));
            if (decoded.length > 0) return decoded;
        }
        return secret.getBytes(StandardCharsets.UTF_8);
    }

    /** Case-insensitively read a single header value. */
    private static String readHeader(Map<String, String> headers, String name) {
        if (headers == null) return null;
        for (Map.Entry<String, String> e : headers.entrySet()) {
            if (e.getKey() != null && e.getKey().equalsIgnoreCase(name)) return e.getValue();
        }
        return null;
    }

    private static boolean isBlank(String s) { return s == null || s.isEmpty(); }
}
